File: Source/ObjectiveC/The.m
//
//  The.m
//  ChangShou
//
//  Created by Mudox on 11/03/2017.
//  Copyright © 2017 Mudox. All rights reserved.
//

#import "The.h"

@implementation The

#pragma mark Commonly accessed global objects

+ (UIDevice *)device {
  return UIDevice.currentDevice;
}

+ (UIApplication *)app {
  return UIApplication.sharedApplication;
}

+ (id<UIApplicationDelegate>)appDelegate {
  return self.app.delegate;
}

+ (UIScreen *)screen {
  return UIScreen.mainScreen;
}

+ (UIWindow *)window {
  return self.app.keyWindow;
}

+ (NSBundle *)bundle {
  return NSBundle.mainBundle;
}

+ (NSFileManager *)fileManager {
  return NSFileManager.defaultManager;
}

+ (NSUserDefaults *)userDefaults {
  return NSUserDefaults.standardUserDefaults;
}

+ (NSNotificationCenter *)notificationCenter {
  return NSNotificationCenter.defaultCenter;
}

#pragma mark Root level view controllers

+ (UIViewController *)rootViewController {
  return self.window.rootViewController;
}

+ (UINavigationController *)rootNavigationController {
  if ([self.rootViewController isKindOfClass:[UINavigationController class]]) {
    return (UINavigationController *)self.rootViewController;
  } else {
    return nil;
  }
}

+ (UITabBarController *)rootTabBarController {
  if ([self.rootViewController isKindOfClass:[UITabBarController class]]) {
    return (UITabBarController *)self.rootViewController;
  } else {
    return nil;
  }
}

#pragma mark Commonly accessed values

+ (CGFloat)screenWidth {
  return CGRectGetWidth(theScreen.bounds);
}

+ (CGFloat)screenHeight {
  return CGRectGetHeight(theScreen.bounds);
}

+ (NSString *)appVersion {
  return [theBundle objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
}

@end

File: MudoxKit/Core/The.swift
//
//  File.swift
//  iOSKit
//
//  Created by Mudox on 21/11/2017.
//

import UIKit

/// The special `enum` used as namespace for all commonly accessed
/// global symbols
public enum The {

  /// UIApplication.shared
  public static var app: UIApplication {
    return UIApplication.shared
  }
  /// UIDevice.current
  public static var device: UIDevice {
    return UIDevice.current
  }

  /// FileManager.default
  public static var fileManager: FileManager {
    return FileManager.default
  }

  /// Bundle.main
  public static var mainBundle: Bundle {
    return Bundle.main
  }

  /// UIScreen.main
  public static var mainScreen: UIScreen {
    return UIScreen.main
  }

  /// UIApplication.shared.windows.first!
  public static var mainWindow: UIWindow {
    return The.app.windows.first!
  }

  /// UIApplication.shared.windows.first!.rootViewController
  public static var rootViewController: UIViewController? {
    return The.mainWindow.rootViewController
  }

  /// NofiticationCenter.default
  public static var notificationCenter: NotificationCenter {
    return NotificationCenter.default
  }

  /// ProcessInfo.processInfo
  public static var process: ProcessInfo {
    return ProcessInfo.processInfo
  }

  /// UserDefaults.standard
  public static var userDefaults: UserDefaults {
    return UserDefaults.standard
  }

}
